Add integrated Metrics support

gitblit opened the pull request:
This is a basic integration of [Metrics](http://metrics.codahale.com) support.  It currently collects and exposes Route and JVM metrics over JMX (install MBean plugin in VisualVM).  Individual controller methods may be annotated with `@Timed` for the integration to collect and report metrics.

This PR will create a merge conflict with #195 since they both implement a `getRoutes()` method for the Router.  Assuming one or the other is merged, I will rebase.

gitblit commented:
Note: There are other things to explore with this integration.  For example, there is an ehCache interface to Metrics - but it is out of date.  It is built against 2.6.6 which has API changes compared to 2.8.x.

buildhive commented:
[Ninja Web Framework » ninja #98](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/98/) UNSTABLE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Ninja Web Framework » ninja #99](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/99/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

gitblit commented:
I backed-off the JVM reporting because of unnecessary unit testing complexity.  It's better, actually, since out-of-the-box it will only report what you register - currently routes.

buildhive commented:
[Ninja Web Framework » ninja #100](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/100/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Ninja Web Framework » ninja #101](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/101/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

raphaelbauer commented:
Hi James,

looks great! You asked in the other PR, why our sketches are added as modules. It's not because of testability. The general consensus is that ninja-core should only contain the core. And be as extensible as possible. That way we get both flexibility for users and don't get too dependent on 3rd party libs. And the release cycle of external modules (your metrics module) are also completely independent of Ninja's release cycle.

In case of your Metrics addition: It would be awesome if you could create a module and push that to maven central. Afaics will work just fine using a NinjaFilter. That way we don't have to add any complexity to ninja-core and still get all benefits of Metrics (+ you are not dependent on Ninja's release cycle for additions/fixes and so on). What do you think?

If there are some missing things (like hooking into router request and so on) let's discuss that and make Ninja as extensible as possible. We are discussing some ideas here and it would be awesome if we could work towards a win-win situation :)

Best,

Raphael

gitblit committed:
Implement a Ninja Metrics module

Applications will need to be manually instrumented by:

1. Adding the optional ninja-metrics dependency
2. Installing the MetricsModule in their conf.Module class
3. Subclassing InstrumentedNinja in their conf.Ninja class
4. Applications may optionally instrument their NinjaCache

Ninja-Metrics will collect data on controller methods annotated
as @Metered or @Timed.  Non-annotated controller methods are
largely ignored except for general count collection.

InstrumentedNinja will monitor bad requests, internal server errors,
and routes not found. It will also keep a real-time count of active
requests.

To try this out:

1. `mvn install` in the repository root
2. `mvn ninja:run` in the ninja-servlet-integration-test
3. Startup VisualVM
4. Install the MBeans plugin through the Tools->Plugins  menu
5. Drill into the NinjaJetty instance listed in VisualVM and select the *MBeans* tab
6. Browse the test app and click on the different available *ninja.cache* and *ninja.requests* metrics

gitblit commented:
I have rewritten and rebased this PR to be based on two new modules:
1. ninja-annotations
2. ninja-metrics

To be completely effective (and agnostic to the underlying http engine), the metrics module needs special consideration in the ninja-core.  For example, I want to automatically wrap Cache instances with InstrumentedCache instances and I want to use InstrumentedNinja instead of DefaultNinja, if it is present and the application does not specify it's own Ninja class.  I also needed to automatically load the Ninja Metrics module so I introduced Reflections and a @NinjaModule annotation to handle that.

This is all subject to change as needed.

To try this out:
1. `mvn install` in the repository root
2. `mvn ninja:run` in the ninja-servlet-integration-test
3. Startup VisualVM
4. Install the MBeans plugin through the Tools->Plugins  menu
5. Drill into the NinjaJetty instance listed in VisualVM and select the _MBeans_ tab
6. Browse the test app and click on the different available _ninja.cache_ and _ninja.requests_ metrics

Example screenshot attached .

![ninja-metrics](https://cloud.githubusercontent.com/assets/868217/3680812/a6b92e3a-12b2-11e4-9797-4efb7a90f905.png)

buildhive commented:
[Ninja Web Framework » ninja #114](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/114/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

gitblit commented:
That CI test failure is a false positive. Everything tested fine.  Failure is something else.

buildhive commented:
[Ninja Web Framework » ninja #166](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/166/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

gitblit commented:
I've rebased this PR on the current HEAD of develop and I have simplified it.  It is now an optional & standlone module that I'd like to see in upstream Ninja instead of sitting around in my repos.  :)
1. I've discarded the changes to core to automatically instrument everything
2. I've discard the stubs for HealthCheck support as I do not have a plan for that at the moment
3. I've updated to the current Metrics release 3.1.0 which has been renamed for DropWizard.

The servlet integration test is instrumented for metrics so if you run that app & startup a tool like VisualVM you can inspect collected metrics.

raphaelbauer commented:
hey gitblit... that looks awesome :) I really want that! I nevertheless wanted to go through the changes in detail and think about some things... Will do within the next week... Thanks for your efforts!

raphaelbauer commented:
Hey gitblit,

some feedback - but nothing to do on your side - I'll propose an implementation the next days:
- tests are missing (not your problem - there were no tests for the cache classes before - that's embarrassing, but I'll change that).
- I'll add graphite support to the core - all our servers are using graphite, so I need that anyway
- I'll try to rewrite the "getRouteMetric" part. afaics we can use guice interceptors to that end. The benefits include a bit cleaner code in the Ninja class - and the ability to measure all methods - not only route methods :)

I'll tackle these 3 things (likely on monday) and get back to you :) Any feedback always welcome! I hope there is nothing I have missed :) Thanks for that great contribution!

Cheers,

Raphael

gitblit commented:
Sounds good.  Looking forward to seeing your version.

raphaelbauer commented:
Just wondering - but is there a special reason why you are using three different registries? Would it be ok to use just one with different prefix?

gitblit commented:
Take a look at the above screenshot of VisualVM.  I wanted to keep the metrics organized.  Cache metrics are discrete and separate from Routing metrics which are discrete and separate from general purpose App metrics - which I would advocate is where your Guice Method Interceptor should log non-controller and non-cache metrics.

Lumping them all together may lead to a mess in sifting through data.  That depends on the JMX client, for sure. I didn't play with JConsole - maybe it's better.

raphaelbauer commented:
ah okay... I see... in graphite it's a bit different...

raphaelbauer committed:
Added interceptors to measure everything (™).

raphaelbauer commented:
Can you have a look at https://github.com/gitblit/ninja/pull/1 (not finished)? Feedback appreciated!

gitblit commented:
Anything you want me to do to move this forward?  Or have I successfully handed this off to you?

raphaelbauer commented:
Actually that was one of the things I wanted to discuss with you... My most important question would be if my PR (gitblit#1) does what you expect it to do - especially because I deleted a lot of your code and replaced it with the two interceptors... If that is ok I'd be happy to finalize everything and testcases and so on...

gitblit commented:
The interceptors are really cool.  They don't measure exactly the same thing.  My version measures the http request from beginning to end - including template/serializer rendering.  Your version measures controller method execution alone.  Both are useful but I guess we can only pick one.

I've been using the end-to-end timing in evaluating the 5 template engines I have.  I have a prototype real-world app and I run the app with each engine over a series of the same pages.  That's pretty useful in evaluating them.

Most people will likely only care about the timings of their specific controller methods - and I probably will too once I choose an engine.

So I think we should move ahead with your interceptor approach.  Slash away.

raphaelbauer commented:
Great..

Btw. You can always annotate the onRoutesRequest method in you conf.Ninja... should work and measure the "whole" request cycle (it still misses everything that happens before, but that's a completely different story).

raphaelbauer commented:
Ah... and much more important... What do the microbenchmarks tell? Which templating engines work well? Which ones do suck?

gitblit commented:
Assuming similar features of the template engines (includes, inheritance, etc) and similar integrations (assetsAt, webJarsAt, i18n, prettyTime), the timing-based benchmarks generally tell me how fast each engine is at rendering the same set of complete pages which use all of the aforementioned features and integrations.  Each of the generated pages also includes file-system retrieved list content so it exercises the template engine's iterator functions as well - 50 items per page.

My findings so far.

 **Dev Mode**
Best performers (all fairly close, but Pebble is the consistent winner):
- Pebble
- FreeMarker
- Trimou

Middle-of-the-road (2x slower than best)
- Thymeleaf

Bottom-of-the-pack (5x slower than best)
- Groovy

**Prod Mode**
Best performers (all fairly close, Pebble & Groovy are top)
- Pebble
- Groovy
- FreeMarker
- Trimou

Middle-of-the-road (2x slower than best)
- Thymeleaf

**The Engines**
[Pebble](http://www.mitchellbosecke.com/pebble/home). String parsing engine Mustache+FreeMarker+AngularJS-like mashup. Nice clean syntax. Young (immature?) engine has at least one important open bug related to if/then/else processing.  I really like this engine- it's the best of the features/capabilities of Mustache & FreeMarker.

[Trimou](http://trimou.org). String parsing engine. Mustache but with server-side extensibility - which you don't get _by design_ from [Mustache.java](https://github.com/spullara/mustache.java).  Developed by a RedHat engineer. Works very well.  Mustache is more readable and understandable than FreeMarker. A solid choice.

[Groovy MarkupTemplateEngine](http://groovy-lang.org/docs/groovy-2.3.7/html/documentation/markup-template-engine.html). Code compiling engine. Surprisingly performant in Prod mode. There are some trade-offs with white-space handling when your templates are code.  It's not a problem, but it may require extra tweaking to get generated pages Just Right.

[Thymeleaf](http://www.thymeleaf.org). DOM parsing engine. All templates must be well-formed XHTML.  All integrations are node attributes.  Disappointing performance, fairly verbose templates, inheritance is a 3rd-party integration.  This would be my last choice and probably not even on my list.

I was also considering trying out [Chunk](http://www.x5software.com/chunk), but I'm not sure I will bother since I think it would be hard to beat Trimou, Pebble, or Groovy.

I'll try to get my modules on Github in the next week or two.

gitblit commented:
Have you had any spare cycles to move metrics forward?

raphaelbauer commented:
Yes! I took a bit longer than expected, but I am working on it right now :)

raphaelbauer committed:
More cleanup for metrics.

raphaelbauer committed:
Some potential code refinments.

raphaelbauer committed:
Some more improvements for Metrics.

raphaelbauer committed:
Fixed problem with wrong source level. Now fixed to 1.7.

gitblit committed:
Reduce to one metric registry, specify ninja.* names, setup JMX with app name

gitblit commented:
Pushed your changes plus simplifications discussed in the other PR.

buildhive commented:
[Ninja Web Framework » ninja #229](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/229/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

gitblit commented:
I have pending updates for one or two of my other proposals but I don't want to overwhelm your free time so I will hold them back until we get Metrics merged. Naturally, I'd like to put this PR to bed sooner than later. :smiley: 

I've taken a look at Graphite and heard it mentioned in a talk at a conference last week.  Sounds useful & interesting. How are you hooking them together?

gitblit commented:
[JMXTrans](https://github.com/jmxtrans/embedded-jmxtrans) perhaps?

gitblit committed:
Use injected MetricRegistry

buildhive commented:
[Ninja Web Framework » ninja #230](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/230/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Ninja Web Framework » ninja #231](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/231/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

gitblit committed:
Add optional JVM and Logback Metrics integration

gitblit committed:
Add optional integrations for Graphite and Ganglia

gitblit commented:
Ok, that was a dumb question.

I've optionally exposed JVM and Logback metrics.  I also added the Graphite and Ganglia reporters.  They are untested as I don't have a working install of either.

I have been playing with Librato (cloud service) which I can say integrates quickly and easily.  Looks like DataDog will also integrate about as easily - I'm sure there are many others too.

buildhive commented:
[Ninja Web Framework » ninja #232](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/232/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

gitblit commented:
I  have also come across what may be a more current alternative to Graphite, InfluxDB+Grafana.

http://influxdb.com
http://grafana.org

Don't know anything about it - but it looks pretty.  ;)

buildhive commented:
[Ninja Web Framework » ninja #233](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/233/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Ninja Web Framework » ninja #234](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/234/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

gitblit committed:
Extracted Ganglia & Graphite reporters into separate jars, added Librato reporter

buildhive commented:
[Ninja Web Framework » ninja #235](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/235/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Ninja Web Framework » ninja #236](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/236/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

gitblit commented:
Alright, I'm done hacking again.  Time to review my changes?

gitblit committed:
Added InfluxDB Metrics reporter

buildhive commented:
[Ninja Web Framework » ninja #237](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/237/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Ninja Web Framework » ninja #238](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/238/) UNSTABLE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

gitblit commented:
@raphaelbauer Do you know why buildhive is running 2x for every pushed change?  It's strange.  BTW the _unstable_ is a false-negative.  The previous triggered build of the same committed code was a success.

I've been playing with InfluxDB+Grafana.  Amazing.  Not sure if you've had the chance to use that yet.  I pushed an integration to this PR last night for it after I said I was done hacking.

gitblit committed:
Added @Counted annotation and matching MethodInterceptor logic

buildhive commented:
[Ninja Web Framework » ninja #240](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/240/) UNSTABLE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Ninja Web Framework » ninja #241](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/241/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

raphaelbauer commented:
Having a look right now :) 
Graphite is great (The classic talk is: https://www.youtube.com/watch?v=czes-oa0yik). Grafana is a nice frontend for Graphite - it's pretty, but I don't know it very well...

raphaelbauer commented:
ah - and "no" - I have no idea why buildhive "double checks" the commits...

gitblit commented:
I've copied the small sources for the influxdb reporter internally (not committed yet) to implement a feature the Librato reporter has - do not report idle stats.  The current reporter dependency will report timers & meters on each pass regardless of whether the metric has changed. Not ideal - although it lets you generate pretty decay graphs.  This change should be upstream - I'll open a PR to the influxdb reporter project after I am satisfied with it.

raphaelbauer commented on ninja-core/src/site/markdown/documentation/basic_concepts/metrics.md:
```diff
@@ -0,0 +1,239 @@
+
+### Reporting Metrics via JMX for VisualVM/JConsole
+
+By default, Ninja Metrics will expose your metrics over JMX. You may disable this behavior by setting *metrics.jmx.enabled=false* in your `application.conf` file.
```
Is that really the default behavior? Imho should be false by default - I think of security and stability issues - or am I overly cautious? (you know JMX better than me apparently...)

raphaelbauer replied:
forget about the comment - I thought all metrics are included within one module - but you created several modules - so when I activate ninja-metrics-graphite - the jmx stuff is not active - right? Therefore it's absolutely okay... :)

gitblit replied:
I don't actually think it enables JMX TCP serving so this is likely poorly worded - and poorly named by the Metrics project.

What I believe it does is exposes the metric registry through MBeans which _can_ be reported via JMX.  MBeans are also visible from VisualVM/JConsole with a direct process connection - which is how I view them.  I don't believe JmxReporter explicitly starts tcp JMX server - I thought that was a VM level function.

We can certainly disable that flag by default and rename it as needed.  Maybe `metrics.mbeans.enabled` ?

raphaelbauer commented on ninja-metrics/src/main/java/ninja/metrics/InstrumentedCache.java:
```diff
@@ -0,0 +1,146 @@
+
+    private final Cache underlyingCache;
+
+    private MetricsService metricsService;
```
we could make all variables final... (of course init() has to be done in the constructor). Tiny thing... Final is always nice...

raphaelbauer commented on ninja-metrics/src/main/java/ninja/metrics/MetricsServiceImpl.java:
```diff
@@ -0,0 +1,189 @@
+        /*
+         * JMX
+         */
+        if (ninjaProps.getBooleanWithDefault("metrics.jmx.enabled", true)) {
```
I'd set that to false by default...

raphaelbauer commented on ninja-metrics/src/test/java/ninja/metrics/InstrumentedCacheTest.java:
```diff
@@ -0,0 +1,49 @@
+    @InjectMocks
+    InstrumentedCache instrumentedCache;
+
+    public void teatAddKeyValueExpiration() {
```
that's a great test! WHO HAS WRITTEN THAT? ...
ops... me... ;)

You can remove my stupid stub if you like...

raphaelbauer commented on ninja-metrics/src/test/java/ninja/metrics/MetricsServiceImplTest.java:
```diff
@@ -0,0 +1,44 @@
+    }
+    
+    @Test
+    public void testSomeMethod() {
```
also a test that could go away...

raphaelbauer commented:
Hey @gitblit 

just had a look at the code... It' really great. Especially as you created individual modules for each reporter... I am wondering if the testcoverage is large enough (there are obviously no integration tests for - say - ganglia). But on the other hand why not it's up to you to decide.

I'd like to put the JMX reporting to false by default. Otherwise I'd merge it (and release Ninja 3.4 asap).

Great work!

Raphael

gitblit committed:
Disable MBeans by default and clarify VisualVM/JConsole/JMX documentation

gitblit committed:
Allow Metrics reporting frequency to be configurable per-reporter

buildhive commented:
[Ninja Web Framework » ninja #242](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/242/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

gitblit commented:
Things I did:
- changed JMX setting name to "metrics.mbeans.enabled", set disabled by default, and updated docs
- added settings per-reporter to control reporting frequency/period
- added a changelog entry
- rebased all changes on develop

Things I did not do:
- remove your test stubs

I believe, other than the lack of tests, this is complete and ready.

buildhive commented:
[Ninja Web Framework » ninja #243](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/243/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

gitblit commented:
I just submitted davidB/metrics-influxdb#10 to skip reporting idle timers, meters, and histograms to InfluxDB.  If that is accepted, merged, and a new artifact is released in a timely fashion, that will definitely be an update I will want to submit to this PR or as a new PR.

As for a next framework release, I'd stay on your existing cadence which puts us at end-of-the-month, right?  We may be able to squeeze in another PR or two.

raphaelbauer commented:
great :) merged!

raphaelbauer commented:
needless to say that your metrics stuff really rocks!

gitblit commented:
_Our_ metrics stuff.  :) Thanks for the help and for merging.
